README.md
# Comments 0.8.2

Simple commenting system.

<p align="center"><img src="comments-screenshot.png?raw=true" alt="Screenshot"></p>

## Installation

1. [Download and install Datenstrom Yellow](https://github.com/datenstrom/yellow/).
2. [Download extension](../../archive/master.zip). If you are using Safari, right click and select 'Download file as'.
4. Copy `comments.zip` into your `system/extensions` folder.

To uninstall delete the [extension files](extension.ini).

## How to use Comments

The extension adds automatically a comments section on blog pages.\*

To add comments on other pages (that is, non-blog pages) create a [comments] shortcut at the end of each page in which you want a comments section. In this case the following optional argument is available:

`opening` = number of days from publication after which comments are closed; this argument, if present, overrides the `commentsOpening` setting (in other words: this value is used and `commentsOpening` is ignored). See below for details and for the meaning of the special values `0` and `-1`.

To add comments on every page of the site, add `<?php echo $this->yellow->page->getExtra("comments") ?>` in  `system/layouts/default.html`, after the line `<?php echo $this->yellow->page->getContent() ?>`.

\* Please remark that in the sample pages of the Blog extension comments are normally closed because of the publication date of the pages!

Once installed, this Comments extension is particularly easy to use:

+ easy for the administrator (zero configuration: all default values are sensible)
+ easy for users (one click for sending a comment, without registration)
+ easy for the moderator (one click in the received email for approving comments)

In spite of that the extension is hard for spammers (spam filter, moderation, honeypot captcha, automatic closing of comments).

## How to configure Comments

The following settings can be configured in file `system/settings/system.ini`. You can leave alone the default values and all will work sensibly.

`commentsModerator` (default: (empty)) = email address of moderator. If not present, main `email` address of site is used; this value can be overridden with a setting `Moderator` in the page.  
`commentsDir` (default:  `comments/`) = the location where your comment files are stored  
`commentsAutoPublish` (default:  `0`) = if set to `1` any comment is published immediately and the moderator can later remove it; if set to `0` no comment is published unless the moderator approves it (except particular cases, this latter behaviour is much more desirable)  
`commentsMaxSize` (default:  `10000`) = maximum size of a comment  
`commentsTimeout` (default:  `0`) = number of days after which a comment is permanently deleted if not approved for publication; if set to `0` comments are never automatically deleted  
`commentsOpening` (default:  `30`) = number of days from publication after which comments are closed; if set to `0` comments are never closed, if set to `-1` all comments are closed regardless of publication date (can be used as a maintenance mode while manually editing the comments file); only `0` and `-1` are meaningful if the metadata of the page does not contain the `published` setting; this value can be overridden with an optional argument when using the shortcut `[comments]`  
`commentsAuthorNotification` (default:  `1`) = if set to `1`, authors are notified with an email of the publication of their comments (useful also as a check on the authenticity of the email entered)  
`commentsSpamFilter` (default:  `href=|url=`) = spam filter as regular expression  
`commentsIconSize` (default:  `80`) = size in pixel of the icon  
`commentsIconGravatar` (default:  `0`) = use [Gravatar](https://en.gravatar.com/) images instead of the internal image creator; when set to `1` also the Name field is syncronously filled in, if available in the Gravatar profile  
`commentsIconGravatarDefault` (default:  `mp`) = default image for Gravatar (see the [documentation](https://en.gravatar.com/site/implement/images/) for possible values); without effect if `commentsIconGravatar` is `0`  

## Example

Embedding comments in a non-blog page:

```
[comments]
[comments 0]
```

## Updating from a previous version

If you were using a version of this extension previous to 0.7.6, you can update the comments to the new format with a simple script:

```
<?php
$extension = "-comments";
$separator = "----";
$dir = "comments/";
$trash = "trash/";

@mkdir($dir);
@mkdir($trash);
$files = glob("*.*");
$count = 0;
foreach ($files as $file) {
    if (preg_match("/^(.*)" . $extension . "(\.\w+)$/", $file, $parts)) {
        $text = str_replace("\r\n", "\n", file_get_contents($file));
        $text = preg_replace("/^". $separator . "$/m", "", $text);
        $new = fopen($dir . $parts[1] . $parts[2], "w") or die("Cannot create new file\n");
        fwrite($new, $text) or die("Cannot write in new file\n");
        fclose($new);
        rename($file, $trash . $file)  or die("Cannot trash old file\n");
        $count++;
    }
}
echo "$count files updated\n";
```

Put it in the blog directory with the name `update.php`, change the first four variables if necessary, and execute with `php update.php`. If all has gone right, delete the script and trash directory. If you cannot easily execute the script remotely, you can for example download all blog files, execute locally the script and then upload again.

## Developers

Previous maintainers nasendackel, wunderfeyd; current maintainer Giovanni Salmeri


CHANGELOG.md
# Changelog 0.8.2

## General

+ Updated to the new API
+ Documentation clarified
+ Language files merged and renamed


# Changelog 0.7.6

## General

+ works with Yellow 0.7.6
+ no more modified snippets for blog and blogpages
+ automatic insertion in blog pages and manual insertion with [comment] shortcode
+ update and micro-optimisation of code

## Settings

+ commentsEmail renamed commentsModerator
+ removed commentsExtension (file mode suppressed)
+ removed commentsTemplate (filename programmatically defined)
+ removed commentsSeparator (superfluous with new parsing)
+ removed commentsAutoAppend (manual appending disallowed)
+ added commentsAuthorNotification 
+ commentsTimeout accepts 0 for no timeout
+ added commentsOpening (number of days, 0 for indefinite, -1 for maintenance mode)
+ removed commentsUrlHighlight (now markdown styling)
+ removed commentsIconBackgroundColor and commentsIconForegroundColors (now hardcoded)
+ commentsIconSize is now in pixels for both types of icons
+ commentsIconGravatarDefault instead of commentsIconGravatarOptions
+ removed commentsBlacklist and blacklist (ineffective, should use DNSBL)

## Metadata

+ moderator setting in page, ovverriding commentsModerator

## Form

+ added checkbox for GDPR conformity
+ markdown styling (PHPMarkdown or Parsedown) in comment area
+ removed website field in form
+ added help messages
+ error messagges and checking uniformed to contact plugin

## Styling

+ synchronous filling in of gravatar and name from gravatar profile
+ auto-growing and -shrinking textarea 
+ css style self-adapting to icon size
+ css highlighting of focussed fields
+ css red colouring of error messages

## Other Safety Features

+ added honeypot captcha
+ IP and timestamp saved in comments file

## Bugs fixed

+ fixed date and hour localisation
+ fixed checking of comment size
+ fixed potential mangling of text in comment


commentscontact.php
<?php

class YellowCommentsContact
{
	const Version = "0.1";
	var $yellow;			//access to API
	
	// Handle initialisation
	function onLoad($yellow)
	{
		$this->yellow = $yellow;
		$this->yellow->config->setDefault("contactSpamFilter", "href=|url=");
	}
	
	// Handle page parsing
	function onParsePage()
	{
		if($this->yellow->page->get("template") == "commentscontact")
		{
			if(PHP_SAPI == "cli") $this->yellow->page->error(500, "Static website not supported!");
			$status = trim($_REQUEST["status"]);
			if($status == "send")
			{
				$status = $this->sendEmail();
				switch($status)
				{
					case "incomplete":	$this->yellow->page->set("contactStatus", $this->yellow->text->get("contactStatusIncomplete")); break;
					case "invalid":		$this->yellow->page->set("contactStatus", $this->yellow->text->get("contactStatusInvalid")); break;
					case "done":		$this->yellow->page->set("contactStatus", $this->yellow->text->get("contactStatusDone")); break;
					case "error":		$this->yellow->page->error(500, $this->yellow->text->get("contactStatusError")); break;
				}
				$this->yellow->page->setHeader("Last-Modified", $this->yellow->toolbox->getHttpDateFormatted(time()));
				$this->yellow->page->setHeader("Cache-Control", "no-cache, must-revalidate");
			} else {
				$status = "none";
				$this->yellow->page->set("contactStatus", $this->yellow->text->get("contactStatusNone"));
			}
			$this->yellow->page->set("status", $status);
		}
	}
	
	// Send contact email
	function sendEmail()
	{
		$status = "send";
		$spamFilter = $this->yellow->config->get("contactSpamFilter");
		if(strempty(trim($_REQUEST["message"]))) $status = "incomplete";
		if(!strempty($_REQUEST["from"]) && !filter_var($_REQUEST["from"], FILTER_VALIDATE_EMAIL)) $status = "invalid";
		if(!strempty($_REQUEST["message"]) && preg_match("/$spamFilter/i", $_REQUEST["message"])) $status = "error";
		$name = preg_replace("/[^\pL\d\-\. ]/u", "-", $_REQUEST["name"]);
		$url = preg_replace("/[^\pL\d\-\. ]/u", "-", $_REQUEST["url"]);
		$beitrag = preg_replace("/[^\pL\d\-\. ]/u", "-", $_REQUEST["beitrag"]);
		$from = preg_replace("/[^\w\-\.\@ ]/", "-", $_REQUEST["from"]);
		if($status == "send")
		{
			$mailMessage = $_REQUEST["message"]."\r\n-- \r\n$name"."\r\n-- \r\n$from"."\r\n-- \r\n$url"."\r\n-- \r\n$beitrag";
			$mailTo = $this->yellow->page->get("contactEmail");
			if($this->yellow->config->isExisting("contactEmail")) $mailTo = $this->yellow->config->get("contactEmail");
			$mailSubject = mb_encode_mimeheader($this->yellow->page->get("title"));
			$mailHeaders = empty($from) ? "From: noreply\r\n" : "From: ".mb_encode_mimeheader($name)." <$from>\r\n";
			$mailHeaders .= "X-Contact-Url: ".mb_encode_mimeheader($this->yellow->page->getUrl())."\r\n";
			$mailHeaders .= "X-Remote-Addr: ".mb_encode_mimeheader($_SERVER["REMOTE_ADDR"])."\r\n";
			$mailHeaders .= "Mime-Version: 1.0\r\n";
			$mailHeaders .= "Content-Type: text/plain; charset=utf-8\r\n";
			$status = mail($mailTo, $mailSubject, $mailMessage, $mailHeaders) ? "done" : "error";
		}
		return $status;
	}
}

$yellow->plugins->register("commentscontact", "YellowCommentsContact", YellowCommentsContact::Version);
?>


comments-gravatar.js
document.addEventListener("DOMContentLoaded", function() {
	document.getElementById('from').addEventListener("change", fillOutFrom);
});
var md5 = function(){for(var m=[],l=0;64>l;)m[l]=0|4294967296*Math.abs(Math.sin(++l));return function(c){var e,g,f,a,h=[];c=unescape(encodeURI(c));for(var b=c.length,k=[e=1732584193,g=-271733879,~e,~g],d=0;d<=b;)h[d>>2]|=(c.charCodeAt(d)||128)<<8*(d++%4);h[c=16*(b+8>>6)+14]=8*b;for(d=0;d<c;d+=16){b=k;for(a=0;64>a;)b=[f=b[3],(e=b[1]|0)+((f=b[0]+[e&(g=b[2])|~e&f,f&e|~f&g,e^g^f,g^(e|~f)][b=a>>4]+(m[a]+(h[[a,5*a+1,3*a+5,7*a][b]%16+d]|0)))<<(b=[7,12,17,22,5,9,14,20,4,11,16,23,6,10,15,21][4*b+a++%4])|f>>>32-b),e,g];for(a=4;a;)k[--a]=k[a]+b[a]}for(c="";32>a;)c+=(k[a>>3]>>4*(1^a++&7)&15).toString(16);return c}}();
function fillOutFrom() {
	var PATT = /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]+$/
	var BASE = "https://en.gravatar.com/";
	var email = this.value.trim().toLowerCase();
	var hash = md5(PATT.test(email) ? email : "");
	var size = document.getElementById("gravatar").width;
	var def = document.getElementById("gravatar").getAttribute("data-default");
	document.getElementById("gravatar").src = BASE+"avatar/"+hash+"?s="+size+"&d="+def;
	var jsonFile = new XMLHttpRequest();
	jsonFile.onreadystatechange = function() {
		if (this.readyState == 4 && this.status == 200) {
			document.getElementById("name").value = JSON.parse(this.responseText).entry[0].displayName;
			document.getElementById("comment").focus();
		} else {
			document.getElementById("name").value ="";
		}
	}
	jsonFile.open("get", BASE+hash+".json", true);
	jsonFile.send();
}


comments.php
<?php
// Comment plugin
// Copyright (c) 2016-2019 Authors (see git commit log)
// This file may be used and distributed under the terms of the public license.

class YellowComment {
    var $metaData;
    var $comment;
    
    // Set comment meta data
    function set($key, $value) {
        $this->metaData[$key] = $value;
    }
    
    // Return comment meta data
    function get($key) {
        return $this->metaData[$key];
    }

    // Return comment meta data, HTML encoded
    function getHtml($key) {
        return htmlspecialchars($this->metaData[$key]);
    }
    
    // Check if comment was published
    function isPublished() {
        return lcfirst($this->get("published")) != "no";
    }
}

class YellowComments {
    const VERSION = "0.8.2";
    const TYPE = "feature";
    public $yellow;         //access to API

    var $comments;
    var $pageText;
    var $fileHandle;
    var $areOpen;

    // Handle initialisation
    public function onLoad($yellow) {
	$this->yellow = $yellow;
        $this->yellow->system->setDefault("commentsModerator", "");
        $this->yellow->system->setDefault("commentsDir", "comments/");
        $this->yellow->system->setDefault("commentsAutoPublish", "0");
        $this->yellow->system->setDefault("commentsMaxSize", "10000");
        $this->yellow->system->setDefault("commentsTimeout", "0");
        $this->yellow->system->setDefault("commentsOpening", "30");
        $this->yellow->system->setDefault("commentsAuthorNotification", "1");
        $this->yellow->system->setDefault("commentsSpamFilter", "href=|url=");
        $this->yellow->system->setDefault("commentsIconSize", "80");
        $this->yellow->system->setDefault("commentsIconGravatar", "0");
        $this->yellow->system->setDefault("commentsIconGravatarDefault", "mp");
    }

    // Handle page content parsing of custom block
    public function onParseContentShortcut($page, $name, $text, $shortcut) {
        $output = null;
        if ($name=="comments" && $shortcut) {
            list($opening) = $this->yellow->toolbox->getTextArgs($text);
            if ($opening == "") $opening = $this->yellow->system->get("commentsOpening");
            $this->areOpen = time()-$opening*86400 < strtotime($this->yellow->page->get("published")) || !$opening;

            $this->lockComments($this->yellow->page, false);
            $this->loadComments();
            $this->processSend();
            if ($this->yellow->page->get("status") == "done") { // post/redirect/get
                setcookie("status", "done");
                $this->yellow->page->clean(303, $this->yellow->page->getLocation(true));
            }
            $this->unlockComments();
            $iconSize = intval($this->yellow->system->get("commentsIconSize")); 

            $output = "<div class=\"comments\" id=\"comments\">\n";
            $output .= "<h2><span>" . $this->yellow->text->get("commentsComments") . " " . $this->getCommentCount() . "</span></h2>\n";
            foreach ((array)$this->comments as $comment) {
                if ($comment->isPublished()) {
                    $output .= "<div class=\"comment\" id=\"" . $comment->getHtml("uid") . "\">\n";
                    $output .= "<div class=\"comment-icon\"><img src=\"" . $this->getUserIcon($comment->get("from")) . "\" width=\"" . $iconSize . "\" height=\"" . $iconSize . "\" alt=\"Image\" /></div>\n";
                    $output .= "<div class=\"comment-main\">\n";
                    $output .= "<div class=\"comment-name\">" . $comment->getHtml("name") . "</div>\n";
                    $output .= "<div class=\"comment-date\">" . $this->yellow->text->normaliseDate($comment->get("created")) . "</div>\n";
                    $output .= "<div class=\"comment-content\">" . $this->transformText($this->yellow->page, $comment->comment) . "</div>\n";
                    $output .= "</div>\n";
                    $output .= "</div>\n";
                }
            }
            $output .= "</div>\n";

//            if ($this->yellow->page->get("status") == "done") {  // works only in FIrefox e Chrome
//                $output .= "<script type=\"text/javascript\">window.addEventListener('load', function() { history.replaceState({}, null, location.href); })</script>\n";
//            }
            if ($_COOKIE["status"] == "done") {
                setcookie("status", "", 1);
                $this->yellow->page->set("status", "done");
            }
            $output .= "<div class=\"content separate\" id=\"form\"></div>\n";
            if ($this->yellow->page->get("status") != "done" && $this->areOpen) {
                $output .= "<p class=\"" . $this->yellow->page->getHtml("status") . "\">" . $this->yellow->text->getHtml("commentsStatus".ucfirst($this->yellow->page->get("status"))) . "</p>\n";
                $output .= "<form class=\"comments-form comment\" action=\"" . $this->yellow->page->getLocation(true) . "#form\" method=\"post\">\n";
                if ($this->yellow->system->get("commentsIconGravatar")) {
                    $output .= "<div class=\"comments-icon\"><img id=\"gravatar\" src=\"" . $this->getUserIcon($this->yellow->page->get("status") == "invalid" ? "" : $_REQUEST["from"]) . "\" width=\"" . $iconSize . "\" height=\"" . $iconSize . "\" data-default=\"" . rawurlencode($this->yellow->system->get("commentsIconGravatarDefault")) . "\" alt=\"Image\" /></div>\n";
                } else {
                    $output .= "<div class=\"comments-icon\"><img src=\"" . $this->getUserIcon($_REQUEST["from"]) . "\" width=\"" . $iconSize . "\" height=\"" . $iconSize . "\" alt=\"Image\" /></div>\n";
                }
                $output .= "<div class=\"comments-main\">\n";
                $output .= "<div class=\"comments-from\"><label for=\"from\">" . $this->yellow->text->getHtml("commentsEmail") . "</label><br /><input type=\"text\" size=\"40\" class=\"form-control\" name=\"from\" id=\"from\" value=\"" . htmlspecialchars($_REQUEST["from"]) . "\" /></div>\n";
                $output .= "<div class=\"comments-name\"><label for=\"name\">" . $this->yellow->text->getHtml("commentsName") . "</label><br /><input type=\"text\" size=\"40\" class=\"form-control\" name=\"name\" id=\"name\" value=\"" . htmlspecialchars($_REQUEST["name"]) . "\" /></div>\n";
                $output .= "<div class=\"comments-message\"><label for=\"message\">" . $this->yellow->text->getHtml("commentsHoneypot") . "</label><br /><textarea class=\"form-control\" name=\"message\" id=\"message\" rows=\"2\" cols=\"70\">" . htmlspecialchars($_REQUEST["message"]) . "</textarea></div>\n";
                $output .= "<div class=\"comments-comment\"><label for=\"comment\">" . $this->yellow->text->getHtml("commentsMessage") . "</label><br /><textarea class=\"form-control\" name=\"comment\" id=\"comment\" rows=\"7\" cols=\"70\" maxlength=\"" . $this->yellow->system->get("commentsMaxSize") . "\">" . htmlspecialchars($_REQUEST["comment"]) . "</textarea></div>\n";
                $output .= "<div class=\"comments-consent\"><input type=\"checkbox\" name=\"consent\" value=\"consent\" id=\"consent\"" . ($_REQUEST["consent"] ? " checked=\"checked\"" : "") . "> <label for=\"consent\">" . $this->yellow->text->getHtml("commentsConsent") . "</label></div>\n";
                $output .= "<div>\n";
                $output .= "<input type=\"hidden\" name=\"status\" value=\"send\" />\n";
                $output .= "<input type=\"submit\" value=\"" . $this->yellow->text->getHtml("commentsButton") . "\" class=\"btn contact-btn\" />\n";
                $output .= "</div>\n";
                $output .= "</div>\n";
                $output .= "</form>\n";
                $output .= "<p class=\"comments-info\">";
                $output .= $this->yellow->text->get("commentsPrivacy") . " ";
                $output .= $this->yellow->system->get("commentsIconGravatar") ? $this->yellow->text->get("commentsGravatar") . " " : "";
                $output .= $this->yellow->text->get("commentsMarkdown") . " ";
                $output .= !$this->yellow->system->get("commentsAutoPublish") ? $this->yellow->text->get("commentsManual") : "";
                $output .= "</p>\n";
            } else {
                $output .= "<p class=\"" . $this->yellow->page->getHtml("status") . "\">" . $this->yellow->text->getHtml("commentsStatus".ucfirst($this->yellow->page->get("status"))) . "</p>\n";
            }
        }
        return $output;
    }

    // Handle page extra data
    public function onParsePageExtra($page, $name) {
        $output = null;
        if ($name=="header") {
            $extensionLocation = $this->yellow->system->get("serverBase").$this->yellow->system->get("extensionLocation");
            $output .= "<link rel=\"stylesheet\" type=\"text/css\" media=\"all\" href=\"{$extensionLocation}comments.css\" />\n";
            $output .= "<script type=\"text/javascript\" defer=\"defer\" src=\"{$extensionLocation}comments-textarea.js\"></script>\n";
            if ($this->yellow->system->get("commentsIconGravatar")) $output .= "<script type=\"text/javascript\" defer=\"defer\" src=\"{$extensionLocation}comments-gravatar.js\"></script>\n";
        }
        if ($name=="comments") {
            $output = $this->onParseContentShortcut($page, "comments", "", true);
        }
        return $output;
    }

    // Return Email
    function getEmail() {
        return $this->yellow->page->get("moderator") ? $this->yellow->page->get("moderator") : ($this->yellow->system->isExisting("commentsModerator") ? $this->yellow->system->get("commentsModerator") : $this->yellow->system->get("email"));
    }

    // Return file name from page object
    function getCommentFileName($page) {
        return dirname($page->fileName) . "/" . $this->yellow->system->get("commentsDir") . basename($page->fileName);
    }

    // Lock comments file
    function lockComments($page, $forceOpen) {
        if ($this->fileHandle)
            return;
        $fileName = $this->getCommentFileName($page);
        if ($forceOpen) @mkdir(dirname($fileName));
        if (file_exists($fileName) || $forceOpen)
            $this->fileHandle = @fopen($fileName, "c+");
        if ($this->fileHandle)
            flock($this->fileHandle, LOCK_EX);
    }

    // Unlock comments file
    function unlockComments() {
        if (!$this->fileHandle)
            return;
        flock($this->fileHandle, LOCK_UN);
        fclose($this->fileHandle);
        unset($this->fileHandle);
    }

    // Load comments from given file name
    function loadComments() {
        if (!$this->fileHandle)
            return;
        fseek($this->fileHandle, 0, SEEK_END);
        $length = ftell($this->fileHandle);
        fseek($this->fileHandle, 0, SEEK_SET);
        $contents = explode("\n\n---\n", str_replace("\r\n", "\n", $length>0 ? fread($this->fileHandle, $length) : ""));
//        more robust (doesn't need \n\n before ---), but slower
//        preg_match_all("/---\n(.+?---\n((?!---\n).)*)/s", ($length>0) ? fread($this->fileHandle, $length) : "", $temp);
//        $contents = $temp[1];
        $pageText = array_shift($contents);
        foreach ($contents as $content) {
            if (preg_match("/^(.+?)\n+---\n+(.*)/s", $content, $parts)) {
                $comment = new YellowComment;
                foreach (preg_split("/\n+/", $parts[1]) as $line) {
                    preg_match("/^\s*(.*?)\s*:\s*(.*?)\s*$/", $line, $matches);
                    if (!empty($matches[1]) && !empty($matches[2])) {
                        $comment->set(lcfirst($matches[1]), $matches[2]);
                        if (lcfirst($matches[1]) == "created") $this->yellow->page->setLastModified(strtotime($matches[2]));
                    }
                }
                $comment->comment = trim($parts[2]);
                $comment->comment = preg_replace("/^-(-{3,})$/m", "$1", $comment->comment); // invert safety substitution
                $this->comments[] = $comment;
            }
        }
    }
    
    // Save comments
    function saveComments($checkSize) {
        if (!$checkSize || (strlenu($comment->comment) < $this->yellow->system->get("commentsMaxSize"))) {
            $status = "send";
            $this->lockComments($this->yellow->page, true);
            if ($this->pageText == "") {
                $this->pageText = @file_get_contents(strreplaceu("(.*)", "comments", $this->yellow->system->get("settingDir").$this->yellow->system->get("newFile")));
                if ($this->pageText == "") { // autogenerate, no need for a template
                    $this->pageText = "---\nTitle: Comments\nParser: comments\n---\n";
                }
            }

            $timeout = time()-$this->yellow->system->get("commentsTimeout")*86400;
            $content = $this->pageText;
            foreach ($this->comments as $comment) {
                if ($comment->isPublished() || $timeout < strtotime($comment->get("created")) || $this->yellow->system->get("commentsTimeout") == 0) {
                    $content .= "\n\n---\n";
                    foreach ($comment->metaData as $key=>$value) {
                        $content .= ucfirst($key). ": " . $value . "\n";
                    }
                    $content .= "---\n";
                    $content .= $comment->comment . "\n";
                }
            }
            if ($this->fileHandle) {
                fseek($this->fileHandle, 0, SEEK_SET);
                fwrite($this->fileHandle, $content);
                ftruncate($this->fileHandle, ftell($this->fileHandle));
            } else {
                $status = "error";
            }
        } else {
            $status = "error";
        }
        return $status;
    }

    // Build comment from input
    function buildComment() {
        $comment = new YellowComment;
        $comment->set("name", filter_var(trim($_REQUEST["name"]), FILTER_SANITIZE_STRING, FILTER_FLAG_STRIP_LOW));
        $comment->set("from", filter_var(trim($_REQUEST["from"]), FILTER_SANITIZE_STRING, FILTER_FLAG_STRIP_LOW));
        $comment->set("created", date("Y-m-d H:i"));
        $comment->set("fingerprint", $_ENV["REMOTE_ADDR"] . "@" . $_SERVER["REQUEST_TIME_FLOAT"]);
        $comment->set("uid", hash("sha256", $this->yellow->toolbox->createSalt(64)));
        $comment->set("aid", hash("sha256", $this->yellow->toolbox->createSalt(64)));
        if (!$this->yellow->system->get("commentsAutoPublish")) $comment->set("published", "No");
        $comment->comment = str_replace("\r\n", "\n", trim($_REQUEST["comment"]));
        $comment->comment = preg_replace("/^-{3,}$/m", "-$0", $comment->comment); // safety substitution
        return $comment;
    }

    // verify comment for safe use
    function verifyComment($comment) {
        $status = "send";
        $name = $comment->get("name");
        $from = $comment->get("from");
        $message = $comment->comment;
        $consent = $_REQUEST["consent"];
        $spamFilter = $this->yellow->system->get("commentsSpamFilter");
        if (empty($name) || empty($from) || empty($message) || empty($consent)) $status = "incomplete";
        if (!empty($from) && !filter_var($from, FILTER_VALIDATE_EMAIL)) $status = "invalid";
        if (!empty($message) && preg_match("/$spamFilter/i", $message)) $status = "error";
        if (!empty($_REQUEST["message"])) $status = "error"; // honeypot
        return $status;
    }

    // Process user input
    function processSend() {
        if ($this->yellow->isCommandLine()) $this->yellow->page->error(500, "Static website not supported!");
        $aid = trim($_REQUEST["aid"]);
        $action = trim($_REQUEST["action"]);
        if ($aid) {
            $changed = false;
            foreach ($this->comments as &$comment) {
                if ($comment->get("aid") == $aid) {
                    if ($action == "remove") {
                        unset($comment);
                        $changed = true;
                        break;
                    } elseif ($action == "publish") {
                        $comment->set("published", null);
                        $changed = true;
                        if ($this->yellow->system->get("commentsAuthorNotification"))
                            $this->sendNotificationEmail($comment);
                        break;
                    }
                }
            }
            if ($changed) $this->saveComments(false);
        }
        $status = trim($_REQUEST["status"]);
        if ($status == "send") {
            $comment = $this->buildComment();
            $status = $this->verifyComment($comment);
            if (!$this->areOpen) $status = "closed";
            if ($status == "send") {
                $this->comments[] = $comment;
                $status = $this->saveComments(true);
            }
            if ($status == "send" && $this->getEmail()) $status = $this->sendEmail($comment);
//            if ($status == "done") { // post/redirect/get
//                setcookie("status", "done");
//                $this->yellow->page->clean(303, $this->yellow->page->getLocation(true));
//            }
            $this->yellow->page->setHeader("Last-Modified", $this->yellow->toolbox->getHttpDateFormatted(time()));
            $this->yellow->page->setHeader("Cache-Control", "no-cache, must-revalidate");
        } else {
            $status = $this->areOpen ? "none" : "closed";
        }
        $this->yellow->page->set("status", $status);
    }
    
    // Send comment email
    function sendEmail($comment) {
        $mailMessage = $comment->comment."\r\n";
        $mailMessage .= "-- \r\n";
        $mailMessage .= "Name: " . $comment->get("name") . "\r\n";
        $mailMessage .= "Mail: " . $comment->get("from") . "\r\n";
        $mailMessage .= "Uid:  " . $comment->get("uid") . "\r\n";
        $mailMessage .= "Fingerprint:  " . $comment->get("fingerprint") . "\r\n";
        $mailMessage .= "-- \r\n";
        if (!$this->yellow->system->get("commentsAutoPublish")) {
            $mailMessage.= "Publish: " . $this->yellow->page->getUrl() . "?aid=" . $comment->get("aid") . "&action=publish\r\n";
        } else {
            $mailMessage.= "Remove: " . $this->yellow->page->getUrl() . "?aid=" . $comment->get("aid") . "&action=remove\r\n";
        }
        $mailSubject = mb_encode_mimeheader("[".$this->yellow->system->get("sitename")."] " . $this->yellow->page->get("title"));
        $mailHeaders = "From: " . mb_encode_mimeheader($comment->get("name")) . " <" . $comment->get("from") . ">\r\n";
        $mailHeaders .= "X-Contact-Url: " . mb_encode_mimeheader($this->yellow->page->getUrl()) . "\r\n";
        $mailHeaders .= "X-Remote-Addr: " . mb_encode_mimeheader($_SERVER["REMOTE_ADDR"]) . "\r\n";
        $mailHeaders .= "Mime-Version: 1.0\r\n";
        $mailHeaders .= "Content-Type: text/plain; charset=utf-8\r\n";
        return mail($this->getEmail(), $mailSubject, $mailMessage, $mailHeaders) ? "done" : "error";
    }

    // Send notification email
    function sendNotificationEmail($comment) {
        $mailMessage = $this->yellow->text->get("commentsPublished")."\r\n\r\n";
        $mailMessage .= $this->yellow->page->getUrl() . "#" . $comment->get("uid") . "\r\n\r\n";
        $mailMessage .= "-- \r\n";
        $mailMessage .= $this->yellow->system->get("sitename") . "\r\n";
        $mailSubject = mb_encode_mimeheader("[".$this->yellow->system->get("sitename")."] " . $this->yellow->page->get("title"));
        $mailHeaders = "From: " . mb_encode_mimeheader($this->yellow->system->get("sitename")) . "<" . $this->yellow->system->get("email") . ">\r\n";
        $mailHeaders .= "Mime-Version: 1.0\r\n";
        $mailHeaders .= "Content-Type: text/plain; charset=utf-8\r\n";
        return mail($comment->get("from"), $mailSubject, $mailMessage, $mailHeaders) ? "done" : "error";
    }

    // Return number of visible comments
    function getCommentCount() {
        $count = 0;
        foreach ((array)$this->comments as $comment) {
            if ($comment->isPublished()) {
                $count++;
            }
        }
        return $count;
    }
    
    // Transform (markdown) text to HTML
    function transformText($page, $text) {
        if (class_exists("MarkdownExtraParser")) {
            $markdownHandler = new MarkdownExtraParser();
            $markdownHandler->no_markup = true;
            $markdownHandler->hard_wrap = true;
            $text = $markdownHandler->transform($text);
        } elseif (class_exists("ParsedownExtra")) {
            $markdownHandler = new ParsedownExtra();
            $markdownHandler->setSafeMode(true);
            $markdownHandler->setBreaksEnabled(true);
            $markdownHandler->setUrlsLinked(true);
            $text = $markdownHandler->text($text);
        }
//        Simpler, but allows shortcodes
//        if ($this->yellow->extensions->isExisting($this->yellow->system->get("parser"))) {
//            $markdownHandler = $this->yellow->extensions->get($this->yellow->system->get("parser"));
//            $page->safeMode = true; // always disallow HTML in comments
//            $text = $markdownHandler->onParseContentRaw($page, $text);
//        }
        $text = preg_replace('/<h\d>(.*)<\/h\d>/', '<p><strong>$1</strong></p>', $text); // no headers, please
        return $text;
    }

    function getUserIcon($email) {
        if ($this->yellow->system->get("commentsIconGravatar")) {
            $base = "//gravatar.com/avatar/";
            return $base . hash("md5", strtolower(trim($email))) . "?s=" . intval($this->yellow->system->get("commentsIconSize")) . "&d=" . rawurlencode($this->yellow->system->get("commentsIconGravatarDefault"));
        } else {
            return "data:image/png;base64," . base64_encode($this->getUserIconPng($email));
        }
    }

    // Get user icon without any service
    function getUserIconPng($email) {
        $hash = hexdec(substr(hash("sha256", strtolower(trim($email))), 0, 6));
        $color_background = 0xFFFFFF;
        $colors = [0xFF0000, 0xCF0000, 0x00FF00, 0x00CF00, 0x0000FF, 0x0000CF, 0xFFCF000, 0xCFFF00, 0x00FFCF, 0x00CFFF, 0xCF00FF, 0xFF00CF];
        $color_foreground = $colors[($hash >> 15) % count($colors)];
        $multiplicator = ceil($this->yellow->system->get("commentsIconSize")/40);
        $size = 5*8*$multiplicator;
        $png = "\x89\x50\x4e\x47\x0d\x0a\x1a\x0a\x00\x00\x00\x0d\x49\x48\x44\x52";
        $png .= pack("N", $size) . pack("N", $size);
        $png .= "\x01\x03\x00\x00\x00";
        $png .= hash("crc32b", substr($png, 0xc), true);
        $png .= "\x00\x00\x00\x06\x50\x4c\x54\x45";
        $png .= substr(pack("N", $color_foreground), 1);
        $png .= substr(pack("N", $color_background), 1);
        $png .= hash("crc32b", substr($png, 0x25), true);
        $map = [0, 1, 2, 1, 0];
        for ($y=0; $y < 5; $y++) {
            $line = "\x00";
            for ($x=0; $x < 5; $x++) {
                $line .= str_repeat(((($hash>>($y*5 + $map[$x])) & 1) == 1) ? "\xff" : "\x00", $multiplicator);
            }
            $pixel .= str_repeat($line, 8*$multiplicator);
        }
        $pixel = gzcompress($pixel, 6);
        $length = strlen($pixel);
        $png .= pack("N", $length);
        $png .= "\x49\x44\x41\x54" . $pixel;
        $png .= hash("crc32b", substr($png, 0x37), true);
        $png .= "\x00\x00\x00\x00\x49\x45\x4e\x44\xae\x42\x60\x82";
        return $png;
    }
}
